File: 백준/Gold/2636. 치즈/치즈.java
import java.util.*;
import java.io.*;

public class Main {
	static int n, m, count = 0;
	static int[][] arr;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		arr = new int[n][m];

		for(int i = 0; i<n; i++){
			st = new StringTokenizer(br.readLine());
			for(int j = 0; j<m; j++){
				arr[i][j] = Integer.parseInt(st.nextToken());
				if(arr[i][j] == 1){
					count++;
				}
			}
		}

		solve(count);
	}
	public static void solve(int cheeseCount){
		int lastCheese = cheeseCount, time = 0;
		while(true){
			time++;
			//찾기
			List<int[]> result = bfs(0, 0);
			cheeseCount -= result.size();
			//만약 없으면 다 녹은거
			if(cheeseCount == 0){
				System.out.print(time +"\n"+lastCheese);
				return;
			}

			//치즈 개수 최신화
			lastCheese = cheeseCount;
			//지우기
			for(int[] pos : result){
				int x = pos[0];
				int y = pos[1];
				arr[x][y] = 0;
			}

		}
	}

	static int[] dx = {-1,1,0,0}, dy = {0,0,-1,1};
	public static List<int[]> bfs(int sx, int sy){
		//찾기
		Queue<int[]> q= new ArrayDeque<>();
		boolean[][] visited = new boolean[n][m];

		q.offer(new int[]{sx, sy});
		visited[sx][sy] = true;

		List<int[]> melt = new ArrayList<>();
		while(!q.isEmpty()){
			int[] cur = q.poll();
			if(arr[cur[0]][cur[1]] == 1) {
				melt.add(new int[]{cur[0], cur[1]});
				continue;
			}
			for(int i = 0; i<4; i++){
				int nx = cur[0] + dx[i];
				int ny = cur[1] + dy[i];

				if(nx<0 || ny<0 || nx>=n || ny>=m|| visited[nx][ny]) continue;
				q.offer(new int[]{nx, ny});
				visited[nx][ny] = true;
			}
		}
		return melt;
	}
}

File: 백준/Gold/2638. 치즈/치즈.kt
import java.util.*

data class Coord(val x: Int, val y: Int)

var N = 0
var M = 0
val arr = Array(100){ IntArray(100) }
val dx = intArrayOf(-1,1,0,0)
val dy = intArrayOf(0,0,-1,1)

fun isValid(x: Int, y: Int): Boolean{
    return (x in 0 until N && y in 0 until M)
}

fun bfs(): Int{
    var time = 0

    while (true) {
        val isVisited = Array(N){ BooleanArray(M) }
        val q = ArrayDeque<Coord>()
        q.add(Coord(0,0))
        isVisited[0][0] = true

        val melt = Array(N){ IntArray(M) }

        // 외부 공기 BFS
        while (q.isNotEmpty()) {
            val cur = q.removeFirst()
            for (i in 0..3) {
                val nx = cur.x + dx[i]
                val ny = cur.y + dy[i]
                if (!isValid(nx, ny) || isVisited[nx][ny]) continue

                if (arr[nx][ny] == 0) {
                    isVisited[nx][ny] = true
                    q.add(Coord(nx, ny))
                } else {
                    melt[nx][ny]++
                }
            }
        }

        var melted = false
        for (i in 0 until N) {
            for (j in 0 until M) {
                if (melt[i][j] >= 2) {
                    arr[i][j] = 0
                    melted = true
                }
            }
        }

        if (!melted) break
        time++
    }

    return time
}

fun main() {
    val br = System.`in`.bufferedReader()
    var st = StringTokenizer(br.readLine())
    N = st.nextToken().toInt()
    M = st.nextToken().toInt()

    for (i in 0 until N) {
        st = StringTokenizer(br.readLine())
        for (j in 0 until M) {
            arr[i][j] = st.nextToken().toInt()
        }
    }

    print(bfs())
}
